tools/fix_bss.py

```diff
@@ -506,27 +506,37 @@ def update_source_file(version_to_update: str, file: Path, new_pragmas: list[Pra
         lines = f.readlines()
 
     for pragma in new_pragmas:
-        line = lines[pragma.line_number - 1]
-        if not line.startswith("#pragma increment_block_number "):
+        i = pragma.line_number - 1
+        if not lines[i].startswith("#pragma increment_block_number "):
             raise FixBssException(
                 f"Expected #pragma increment_block_number on line {pragma.line_number}"
             )
 
-        # Grab pragma argument and remove quotes
-        arg = line.strip()[len("#pragma increment_block_number ") + 1 : -1]
-
-        amounts_by_version = {}
-        for part in arg.split():
-            version, amount_str = part.split(":")
-            amounts_by_version[version] = int(amount_str)
-
-        amounts_by_version[version_to_update] = pragma.amount
-        new_arg = " ".join(
-            f"{version}:{amount}" for version, amount in sorted(amounts_by_version.items())
-        )
-        new_line = f'#pragma increment_block_number "{new_arg}"\n'
-
-        lines[pragma.line_number - 1] = new_line
+        # list the pragma line and any continuation line
+        pragma_lines = [lines[i]]
+        while pragma_lines[-1].endswith("\\\n"):
+            i += 1
+            pragma_lines.append(lines[i])
+
+        for i, s in enumerate(pragma_lines, pragma.line_number - 1):
+            s_updated, n = re.subn(
+                rf"{version_to_update}:(\d+)\b",
+                f"{version_to_update}:{pragma.amount}",
+                s,
+            )
+            if n != 0:
+                # version found and amount updated, set the modified line
+                lines[i] = s_updated
+                break
+        else:
+            # the version is not in the existing pragma
+            i = pragma.line_number - 1
+            # add version:amount between the pragma prefix and the other arguments
```

Comment:
in my PR I changed this to sort the versions because I didn't like how they were arbitrarily ordered by whenever you happened to ran the tool. Is this worth worrying about?

Reply:
this now conserves the existing order, so when a version is added you merely have to manually move it where you want
adding the version anywhere else than at the start is tougher to do

Reply:
Here's one way:

```python
# Parses #pragma increment_block_number (with line continuations already removed)
def parse_pragma(pragma_string: str) -> dict[str, int]:
    amounts = {}
    for part in pragma_string.replace('"', '').split()[2:]:
        kv = part.split(":")
        if len(kv) != 2:
            raise Exception("#pragma increment_block_number arguments followed by a list of version:amount pairs")
        try:
            amount = int(kv[1])
        except ValueError:
            raise Exception("#pragma increment_block_number amount must be an integer")
        amounts[kv[0]] = amount
    return amounts

# Formats #pragma increment_block_number as a list of lines
def format_pragma(amounts: dict[str, int], max_line_length: int) -> list[str]:
    lines = []
    pragma_start = "#pragma increment_block_number "
    current_line = pragma_start + '"'
    first = True
    for version, amount in sorted(amounts.items()):
        part = f"{version}:{amount}"
        if len(current_line) + len(part) > max_line_length - 3:
            lines.append(current_line + '" \\\n')
            current_line = " " * len(pragma_start) + '"'
            first = True
        if not first:
            current_line += " "
        current_line += part
        first = False
    lines.append(current_line + '"\n')
    return lines

if __name__ == "__main__":
    pragma_string = '#pragma increment_block_number "gc-eu:64 gc-eu-mq:64 gc-eu-mq-dbg:0" '
    pragma_string += '    "gc-jp:64 gc-jp-ce:64 gc-jp-mq:64" '
    pragma_string += '    "gc-us:64 gc-us-mq:64"'

    amounts = parse_pragma(pragma_string)
    print(amounts)
    # {'gc-eu': 64, 'gc-eu-mq': 64, 'gc-eu-mq-dbg': 0, 'gc-jp': 64, 'gc-jp-ce': 64, 'gc-jp-mq': 64, 'gc-us': 64, 'gc-us-mq': 64}

    for line in format_pragma(amounts, 80):
        print(line, end='')
    # #pragma increment_block_number "gc-eu:64 gc-eu-mq:64 gc-eu-mq-dbg:0 gc-jp:64" \
    #                                "gc-jp-ce:64 gc-jp-mq:64 gc-us:64 gc-us-mq:64"
```

Reply:
sure, feel free to integrate this

Reply:
alright, I'll PR this after the GC retail PR is in